### DISCUSSION TITLE
Using navigationDestination on a view with List running a task or onAppear doesn't trigger view update on state change

### DISCUSSION THREAD

**Discussion by kakhaberikiknadze** [pointfreeco/swift-composable-architecture#1814] [Q&A]
### Description

# What/Why
`WithViewStore` `body` is not re-triggered when Reducer changes the state

# Screen Recording

https://user-images.githubusercontent.com/79094816/211607313-79754597-9b2c-4009-83d1-40ec8a0c5bee.mp4

As you can see in the video footage it works in case of using `NavigationLink` with `(destination: label:)` init but the state is not updated in case of `(value: label:)` initialiser. 

## Note
View will be updated if you add a button that sends the same action and tap it.
The same setup would work if you remove `List`

### Checklist

- [ ] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [ ] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Expected behavior

Both - "Navigation link destination" and "Navigation destination" should be showing two texts when opened (Initial text & new text).

### Actual behavior

Only "Navigation link destination" shows the correct view state (initial text & new text).

### Steps to reproduce

Can be reproduced using this simple implementation.

``` Swift
import SwiftUI
import ComposableArchitecture

func getAsyncText() async -> String {
    return "new text"
}

struct Reducer: ReducerProtocol {
    struct State: Hashable {
        var texts: [String] = ["Initial text"]
    }
    
    enum Action {
        case changeText
        case changeTextResponse(TaskResult<String>)
    }
    
    var body: some ReducerProtocol<State, Action> {
        Reduce { state, action in
            switch action {
            case .changeText:
                return .task {
                    await .changeTextResponse(
                        TaskResult(catching: {
                            await getAsyncText()
                        })
                    )
                }
            case let .changeTextResponse(.success(newText)):
                state.texts.append(newText)
                return .none
            case .changeTextResponse(.failure):
                return .none
            }
        }
    }
}

struct NavigationDestinationIssueView: View {
    let store: StoreOf<Reducer>
    
    var body: some View {
        WithViewStore(store, observe: { $0 }) { viewStore in
            List(viewStore.texts, id: \.self) { text in
                Text(text)
            }
            .task {
                viewStore.send(.changeText)
            }
        }
    }
}
```

Then try adding `NavigationStack` in your `ContentView` and add two options of `NavigationLink` - One initialised with `destination` and the other with `value`.

``` Swift
struct ContentView: View {
    var body: some View {
        NavigationStack {
            List {
                NavigationLink {
                    NavigationDestinationIssueView(store: .init(
                        initialState: Reducer.State(),
                        reducer: Reducer()
                    ))
                } label: {
                    Text("Navigation link destination")
                }
                
                NavigationLink(value: Reducer.State()) {
                    Text("Navigation destination")
                }

            }
            .navigationTitle("Home")
            // FAIL: - Changing view state from .task or .onAppear doesn't trigger WithViewStore content closure
            .navigationDestination(for: Reducer.State.self) { initialState in
                NavigationDestinationIssueView(store: .init(
                    initialState: initialState,
                    reducer: Reducer()
                ))
            }
        }
    }
}
```

### The Composable Architecture version information

0.48.1

### Destination operating system

iOS 16

### Xcode version information

Version 14.0.1 (14A400)

### Swift Compiler version information

```shell
swift-driver version: 1.62.8 Apple Swift version 5.7 (swiftlang-5.7.0.127.4 clang-1400.0.29.50)
Target: arm64-apple-macosx13.0
```

**Comment by mbrandonw** [comment]
Hi @kakhaberikiknadze, this is happening because you are constructing all new stores every time navigation happens (and there's another issue with data in the array having duplicate ids which is a no-no in `ForEach`).

Typically a store is constructed from scratch only a single time, at the root of the application. Then you use `scope` to derive stores for child domains to pass to child views.

In your example, you can just construct the store once in `ContentView` and then pass it around:

```swift
struct ContentView: View {
  let store = Store(
    initialState: Reducer.State(),
    reducer: Reducer()
  )
  var body: some View {
    NavigationStack {
      List {
        NavigationLink {
          NavigationDestinationIssueView(store: store)
        } label: {
          Text("Navigation link destination")
        }

        NavigationLink(value: Reducer.State()) {
          Text("Navigation destination")
        }

      }
      .navigationTitle("Home")
      // FAIL: - Changing view state from .task or .onAppear doesn't trigger WithViewStore content closure
      .navigationDestination(for: Reducer.State.self) { initialState in
        NavigationDestinationIssueView(store: store)
      }
    }
  }
}
```

I also recommend making sure that the elements in the array each have unique identity. This is one way to do that for your demo:

```swift
func getAsyncText() async -> String {
  return "new text \(Int.random(in: 1...1_000_000_000))"
}
```

But really you should make use of `IdentifiedArray` for that kind of stuff.

Also since this is not an issue with the library I am going to convert this to a discussion.

**Reply by kakhaberikiknadze** [reply]
Thanks for the answer 🙌🏻 I got your point. As I see there's also no more need to use `Reducer.State()` as the destination value, right?

Btw, is there a case-study for `NavigationStack`? I couldn't find it. Would be great to see its proper use cases.

**Reply by stephencelis** [reply]
@kakhaberikiknadze We don't yet ship more navigation-specific tools for TCA, but have been actively working on refining solutions to the problem for awhile and is the next focus for the library :smile: You can read more about it here:

https://github.com/pointfreeco/swift-composable-architecture/discussions/1477
